{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "50028e51-d24f-4ced-9827-5a4f4f04b4e7",
   "metadata": {},
   "source": [
    "# Prototype of associating variant annotations with clinical annotations\n",
    "\n",
    "Goal:\n",
    "* For each clinical annotation genotype/haplotype ID, associate one or more variant annotation IDs\n",
    "* Try a simple algorithm first, check approx. what percentage we can associate naively\n",
    "\n",
    "Known issues:\n",
    "* What to do about ref/ref genotypes?\n",
    "    * Curious how often the \"comparison allele\" is the reference?\n",
    "        * not always, e.g. [here](https://www.pharmgkb.org/variantAnnotation/1451665780)\n",
    "* How to go from statement about alleles to a statement about genotypes?\n",
    "    * Or vice versa, apparently\n",
    "* What to do about \"metabolyzer\" types?\n",
    "\n",
    "## Table of contents\n",
    "\n",
    "1. [Intro](#Intro)\n",
    "    1. [Example 1](#Example-1)\n",
    "    2. [Example 2](#Example-2)\n",
    "2. [Algorithm](#Algorithm)\n",
    "    1. [Prototyping](#Prototyping)\n",
    "    2. [Full dataset run](#Full-dataset-run)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eacb8517-0f93-4c92-aff3-2bafbeeae0ed",
   "metadata": {},
   "source": [
    "## Intro\n",
    "\n",
    "[Top of page](#Table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b00e6812-ed83-412f-9427-cd9af9d39b44",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import re\n",
    "from collections import Counter\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from opentargets_pharmgkb.evidence_generation import ID_COL_NAME\n",
    "from opentargets_pharmgkb.pandas_utils import read_tsv_to_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9234a3a0-2ba0-40d4-99b1-a7c1d92ae0e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 100)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d49dad9a-6e21-40d3-9e32-a7821e64400d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/home/april/projects/opentargets/pharmgkb/doe'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d52e67a8-4c8d-434c-8b5d-d8a8d628281c",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_drug_ann = read_tsv_to_df(os.path.join(data_dir, 'var_drug_ann.tsv'))\n",
    "var_fa_ann = read_tsv_to_df(os.path.join(data_dir, 'var_fa_ann.tsv'))\n",
    "var_pheno_ann = read_tsv_to_df(os.path.join(data_dir, 'var_pheno_ann.tsv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e92193b4-20ac-4865-b6ee-9ecedb1ac175",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinical_annotations = read_tsv_to_df(os.path.join(data_dir, 'clinical_annotations.tsv'))\n",
    "clinical_ann_evidence = read_tsv_to_df(os.path.join(data_dir, 'clinical_ann_evidence.tsv'))\n",
    "clinical_ann_alleles = read_tsv_to_df(os.path.join(data_dir, 'clinical_ann_alleles.tsv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51c72b06-e8e7-48c3-b5d3-9f3ccdbea3db",
   "metadata": {},
   "source": [
    "### Example 1\n",
    "\n",
    "[Top of page](#Table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bdb56ccb-627c-40a8-be41-0fd9d93841ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Specialty Population</th>\n",
       "      <th>Metabolizer types</th>\n",
       "      <th>isPlural</th>\n",
       "      <th>Is/Is Not associated</th>\n",
       "      <th>Direction of effect</th>\n",
       "      <th>PD/PK terms</th>\n",
       "      <th>Multiple drugs And/or</th>\n",
       "      <th>Population types</th>\n",
       "      <th>Population Phenotypes or diseases</th>\n",
       "      <th>Multiple phenotypes or diseases And/or</th>\n",
       "      <th>Comparison Allele(s) or Genotype(s)</th>\n",
       "      <th>Comparison Metabolizer types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1448997750</td>\n",
       "      <td>CYP2B6*1, CYP2B6*18</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>16495778</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that in the paper the allele was referred to as CYP2B6*16. CYP2B6*16 and *18 alleles have been consolidated by PharmVar in Jan 2020, with *16 now listed as a suballele of *18 (CYP2B6*18.002). This annotation is updated to be on CYP2B6*18, instead of CYP2B6*16.</td>\n",
       "      <td>CYP2B6 *1/*18 is associated with increased concentrations of efavirenz in people with HIV Infections as compared to CYP2B6 *1/*1.</td>\n",
       "      <td>*1/*18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Is</td>\n",
       "      <td>Associated with</td>\n",
       "      <td>increased</td>\n",
       "      <td>concentrations of</td>\n",
       "      <td>NaN</td>\n",
       "      <td>in people with</td>\n",
       "      <td>Disease:HIV Infections</td>\n",
       "      <td>NaN</td>\n",
       "      <td>*1/*1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Variant Annotation ID   Variant/Haplotypes    Gene    Drug(s)      PMID  \\\n",
       "3            1448997750  CYP2B6*1, CYP2B6*18  CYP2B6  efavirenz  16495778   \n",
       "\n",
       "  Phenotype Category Significance  \\\n",
       "3      Metabolism/PK          yes   \n",
       "\n",
       "                                                                                                                                                                                                                                                                              Notes  \\\n",
       "3  Please note that in the paper the allele was referred to as CYP2B6*16. CYP2B6*16 and *18 alleles have been consolidated by PharmVar in Jan 2020, with *16 now listed as a suballele of *18 (CYP2B6*18.002). This annotation is updated to be on CYP2B6*18, instead of CYP2B6*16.   \n",
       "\n",
       "                                                                                                                            Sentence  \\\n",
       "3  CYP2B6 *1/*18 is associated with increased concentrations of efavirenz in people with HIV Infections as compared to CYP2B6 *1/*1.   \n",
       "\n",
       "  Alleles Specialty Population Metabolizer types isPlural  \\\n",
       "3  *1/*18                  NaN               NaN       Is   \n",
       "\n",
       "  Is/Is Not associated Direction of effect        PD/PK terms  \\\n",
       "3      Associated with           increased  concentrations of   \n",
       "\n",
       "  Multiple drugs And/or Population types Population Phenotypes or diseases  \\\n",
       "3                   NaN   in people with            Disease:HIV Infections   \n",
       "\n",
       "  Multiple phenotypes or diseases And/or Comparison Allele(s) or Genotype(s)  \\\n",
       "3                                    NaN                               *1/*1   \n",
       "\n",
       "  Comparison Metabolizer types  \n",
       "3                          NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_drug_ann[var_drug_ann['Variant Annotation ID'] == '1448997750']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e2feb058-1bdd-440e-8ff9-1dc8569598cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Specialty Population</th>\n",
       "      <th>Metabolizer types</th>\n",
       "      <th>isPlural</th>\n",
       "      <th>Is/Is Not associated</th>\n",
       "      <th>Direction of effect</th>\n",
       "      <th>PD/PK terms</th>\n",
       "      <th>Multiple drugs And/or</th>\n",
       "      <th>Population types</th>\n",
       "      <th>Population Phenotypes or diseases</th>\n",
       "      <th>Multiple phenotypes or diseases And/or</th>\n",
       "      <th>Comparison Allele(s) or Genotype(s)</th>\n",
       "      <th>Comparison Metabolizer types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>1184988315</td>\n",
       "      <td>rs28399499</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>20952418</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>*Note: combined analysis with rs3745274* Mean log efavirenz trough concentrations increased with the number of rs3745274 T or rs28399499 C alleles.</td>\n",
       "      <td>Allele C is associated with decreased metabolism of efavirenz in people with HIV Infections.</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Is</td>\n",
       "      <td>Associated with</td>\n",
       "      <td>decreased</td>\n",
       "      <td>metabolism of</td>\n",
       "      <td>NaN</td>\n",
       "      <td>in people with</td>\n",
       "      <td>Disease:HIV Infections</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant Annotation ID Variant/Haplotypes    Gene    Drug(s)      PMID  \\\n",
       "299            1184988315         rs28399499  CYP2B6  efavirenz  20952418   \n",
       "\n",
       "    Phenotype Category Significance  \\\n",
       "299      Metabolism/PK          yes   \n",
       "\n",
       "                                                                                                                                                   Notes  \\\n",
       "299  *Note: combined analysis with rs3745274* Mean log efavirenz trough concentrations increased with the number of rs3745274 T or rs28399499 C alleles.   \n",
       "\n",
       "                                                                                         Sentence  \\\n",
       "299  Allele C is associated with decreased metabolism of efavirenz in people with HIV Infections.   \n",
       "\n",
       "    Alleles Specialty Population Metabolizer types isPlural  \\\n",
       "299       C                  NaN               NaN       Is   \n",
       "\n",
       "    Is/Is Not associated Direction of effect    PD/PK terms  \\\n",
       "299      Associated with           decreased  metabolism of   \n",
       "\n",
       "    Multiple drugs And/or Population types Population Phenotypes or diseases  \\\n",
       "299                   NaN   in people with            Disease:HIV Infections   \n",
       "\n",
       "    Multiple phenotypes or diseases And/or  \\\n",
       "299                                    NaN   \n",
       "\n",
       "    Comparison Allele(s) or Genotype(s) Comparison Metabolizer types  \n",
       "299                                 NaN                          NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_drug_ann[var_drug_ann['Variant Annotation ID'] == '1184988315']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2515bed9-3de2-40f7-b983-c019d514605d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Evidence ID</th>\n",
       "      <th>Evidence Type</th>\n",
       "      <th>Evidence URL</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Summary</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13356</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>1184988315</td>\n",
       "      <td>Variant Drug Annotation</td>\n",
       "      <td>https://www.pharmgkb.org/variantAnnotation/1184988315</td>\n",
       "      <td>20952418</td>\n",
       "      <td>Allele C is associated with decreased metabolism of efavirenz in people with HIV Infections.</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Clinical Annotation ID Evidence ID            Evidence Type  \\\n",
       "13356             1184133833  1184988315  Variant Drug Annotation   \n",
       "\n",
       "                                                Evidence URL      PMID  \\\n",
       "13356  https://www.pharmgkb.org/variantAnnotation/1184988315  20952418   \n",
       "\n",
       "                                                                                            Summary  \\\n",
       "13356  Allele C is associated with decreased metabolism of efavirenz in people with HIV Infections.   \n",
       "\n",
       "      Score  \n",
       "13356   1.5  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_ann_evidence[clinical_ann_evidence['Evidence ID'] == '1184988315']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2fd5bc63-61a2-4394-96ad-ab1e0b21375f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Level of Evidence</th>\n",
       "      <th>Level Override</th>\n",
       "      <th>Level Modifiers</th>\n",
       "      <th>Score</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>PMID Count</th>\n",
       "      <th>Evidence Count</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>Phenotype(s)</th>\n",
       "      <th>Latest History Date (YYYY-MM-DD)</th>\n",
       "      <th>URL</th>\n",
       "      <th>Specialty Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4493</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>CYP2B6*1, CYP2B6*4, CYP2B6*6, CYP2B6*9, CYP2B6*18, CYP2B6*28</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Tier 1 VIP</td>\n",
       "      <td>361.5</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>44</td>\n",
       "      <td>69</td>\n",
       "      <td>efavirenz</td>\n",
       "      <td>HIV Infections</td>\n",
       "      <td>2021-03-24</td>\n",
       "      <td>https://www.pharmgkb.org/clinicalAnnotation/1184133833</td>\n",
       "      <td>Pediatric</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Clinical Annotation ID  \\\n",
       "4493             1184133833   \n",
       "\n",
       "                                                Variant/Haplotypes    Gene  \\\n",
       "4493  CYP2B6*1, CYP2B6*4, CYP2B6*6, CYP2B6*9, CYP2B6*18, CYP2B6*28  CYP2B6   \n",
       "\n",
       "     Level of Evidence Level Override Level Modifiers  Score  \\\n",
       "4493                1A            NaN      Tier 1 VIP  361.5   \n",
       "\n",
       "     Phenotype Category PMID Count Evidence Count    Drug(s)    Phenotype(s)  \\\n",
       "4493      Metabolism/PK         44             69  efavirenz  HIV Infections   \n",
       "\n",
       "     Latest History Date (YYYY-MM-DD)  \\\n",
       "4493                       2021-03-24   \n",
       "\n",
       "                                                         URL  \\\n",
       "4493  https://www.pharmgkb.org/clinicalAnnotation/1184133833   \n",
       "\n",
       "     Specialty Population  \n",
       "4493            Pediatric  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_annotations[clinical_annotations[ID_COL_NAME] == '1184133833']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eeefac75-6161-42ea-9340-e631ba5a0f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Genotype/Allele</th>\n",
       "      <th>Annotation Text</th>\n",
       "      <th>Allele Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13707</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*1</td>\n",
       "      <td>The CYP2B6*1 allele is assigned as a normal function allele by CPIC. Patients carrying CYP2B6*1 allele in combination with another normal function allele may have increased metabolism and decreased concentrations of efavirenz as compared to patients with a no or decreased function allele in combination with a normal or increased function allele or with two no or decreased function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>Normal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13708</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*4</td>\n",
       "      <td>The CYP2B6*4 allele is assigned as an increased function allele by CPIC. Patients carrying the CYP2B6*4 allele in combination with a normal function allele or another increased function allele may have increased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>Increased function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13709</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*6</td>\n",
       "      <td>The CYP2B6*6 allele is assigned as a decreased function allele by CPIC. Patients carrying the CYP2B6*6 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>Decreased function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13710</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*9</td>\n",
       "      <td>The CYP2B6*9 allele is assigned as a decreased function allele by CPIC. Patients carrying the CYP2B6*9 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>Decreased function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13711</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*18</td>\n",
       "      <td>The CYP2B6*18 allele is assigned as a no function allele by CPIC. Patients carrying the CYP2B6*18 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>No function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13712</th>\n",
       "      <td>1184133833</td>\n",
       "      <td>*28</td>\n",
       "      <td>The CYP2B6*18 allele is assigned as a no function allele by CPIC. Patients carrying the CYP2B6*28 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.</td>\n",
       "      <td>No function</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Clinical Annotation ID Genotype/Allele  \\\n",
       "13707             1184133833              *1   \n",
       "13708             1184133833              *4   \n",
       "13709             1184133833              *6   \n",
       "13710             1184133833              *9   \n",
       "13711             1184133833             *18   \n",
       "13712             1184133833             *28   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 Annotation Text  \\\n",
       "13707  The CYP2B6*1 allele is assigned as a normal function allele by CPIC. Patients carrying CYP2B6*1 allele in combination with another normal function allele may have increased metabolism and decreased concentrations of efavirenz as compared to patients with a no or decreased function allele in combination with a normal or increased function allele or with two no or decreased function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "13708                                                                                                     The CYP2B6*4 allele is assigned as an increased function allele by CPIC. Patients carrying the CYP2B6*4 allele in combination with a normal function allele or another increased function allele may have increased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "13709                                                                                                              The CYP2B6*6 allele is assigned as a decreased function allele by CPIC. Patients carrying the CYP2B6*6 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "13710                                                                                                                                                               The CYP2B6*9 allele is assigned as a decreased function allele by CPIC. Patients carrying the CYP2B6*9 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "13711                                                                                                                   The CYP2B6*18 allele is assigned as a no function allele by CPIC. Patients carrying the CYP2B6*18 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. However, conflicting evidence has been reported. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "13712                                                                                                                                                                    The CYP2B6*18 allele is assigned as a no function allele by CPIC. Patients carrying the CYP2B6*28 allele in combination with a no, decreased, normal, or increased function allele may have decreased metabolism of efavirenz as compared to patients with two normal function alleles. Other genetic and clinical factors may also influence metabolism of efavirenz. This annotation only covers the pharmacokinetic relationship between CYP2B6 and efavirenz and does not include evidence about clinical outcomes.   \n",
       "\n",
       "          Allele Function  \n",
       "13707     Normal function  \n",
       "13708  Increased function  \n",
       "13709  Decreased function  \n",
       "13710  Decreased function  \n",
       "13711         No function  \n",
       "13712         No function  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_ann_alleles[clinical_ann_alleles[ID_COL_NAME] == '1184133833']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "527681da-7bea-4761-be96-45340fbc54bb",
   "metadata": {},
   "source": [
    "Variant annotation sentence:\n",
    "```\n",
    "CYP2B6 *1/*18 is associated with increased concentrations of efavirenz in people with HIV Infections as compared to CYP2B6 *1/*1.\n",
    "```\n",
    "\n",
    "Clinical annotation has sentences for `*1` & `*18`, need to associate with one or both of them.\n",
    "\n",
    "It's actually much worse - there are variant annotations associated with this clinical annotation that refer to alleles of one of the RS composing the named allele!\n",
    "\n",
    "[Here](https://www.pharmgkb.org/variantAnnotation/1184988315) is an example. Variant annotation sentence:\n",
    "```\n",
    "Allele C is associated with decreased metabolism of efavirenz in people with HIV Infections.\n",
    "```\n",
    "This is for a specific variant rs28399499. This would have to be associated with one or more of the star alleles of CYP2B6 based on the allele definition table (in this case, only `*18` has C for that RS).\n",
    "\n",
    "Also note the direction flipping - \"increased concentration\" to \"decreased metabolism\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3069c06-f757-476f-88f2-427218a81004",
   "metadata": {},
   "source": [
    "### Example 2\n",
    "\n",
    "[Top of page](#Table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7249a975-b0ee-4e48-8531-a25a03bd9bd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Variant/Haplotypes</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Drug(s)</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>Specialty Population</th>\n",
       "      <th>Metabolizer types</th>\n",
       "      <th>isPlural</th>\n",
       "      <th>Is/Is Not associated</th>\n",
       "      <th>Direction of effect</th>\n",
       "      <th>PD/PK terms</th>\n",
       "      <th>Multiple drugs And/or</th>\n",
       "      <th>Population types</th>\n",
       "      <th>Population Phenotypes or diseases</th>\n",
       "      <th>Multiple phenotypes or diseases And/or</th>\n",
       "      <th>Comparison Allele(s) or Genotype(s)</th>\n",
       "      <th>Comparison Metabolizer types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7365</th>\n",
       "      <td>981755665</td>\n",
       "      <td>rs75527207</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>ivacaftor</td>\n",
       "      <td>21083385</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>not stated</td>\n",
       "      <td>Clinical trials were carried out to test efficacy of ivacaftor selecting only patients with the CFTR G551D mutation on at least one allele (genotype AA or AG).</td>\n",
       "      <td>Genotypes AA + AG are associated with response to ivacaftor in people with Cystic Fibrosis.</td>\n",
       "      <td>AA + AG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Are</td>\n",
       "      <td>Associated with</td>\n",
       "      <td>NaN</td>\n",
       "      <td>response to</td>\n",
       "      <td>NaN</td>\n",
       "      <td>in people with</td>\n",
       "      <td>Disease:Cystic Fibrosis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Variant Annotation ID Variant/Haplotypes  Gene    Drug(s)      PMID  \\\n",
       "7365             981755665         rs75527207  CFTR  ivacaftor  21083385   \n",
       "\n",
       "     Phenotype Category Significance  \\\n",
       "7365           Efficacy   not stated   \n",
       "\n",
       "                                                                                                                                                                Notes  \\\n",
       "7365  Clinical trials were carried out to test efficacy of ivacaftor selecting only patients with the CFTR G551D mutation on at least one allele (genotype AA or AG).   \n",
       "\n",
       "                                                                                         Sentence  \\\n",
       "7365  Genotypes AA + AG are associated with response to ivacaftor in people with Cystic Fibrosis.   \n",
       "\n",
       "      Alleles Specialty Population Metabolizer types isPlural  \\\n",
       "7365  AA + AG                  NaN               NaN      Are   \n",
       "\n",
       "     Is/Is Not associated Direction of effect  PD/PK terms  \\\n",
       "7365      Associated with                 NaN  response to   \n",
       "\n",
       "     Multiple drugs And/or Population types Population Phenotypes or diseases  \\\n",
       "7365                   NaN   in people with           Disease:Cystic Fibrosis   \n",
       "\n",
       "     Multiple phenotypes or diseases And/or  \\\n",
       "7365                                    NaN   \n",
       "\n",
       "     Comparison Allele(s) or Genotype(s) Comparison Metabolizer types  \n",
       "7365                                 NaN                          NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_drug_ann[var_drug_ann['Variant Annotation ID'] == '981755665']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea4d211e-2b05-48f2-b1a1-d328bfb1b1b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Evidence ID</th>\n",
       "      <th>Evidence Type</th>\n",
       "      <th>Evidence URL</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Summary</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>981755803</td>\n",
       "      <td>981755665</td>\n",
       "      <td>Variant Drug Annotation</td>\n",
       "      <td>https://www.pharmgkb.org/variantAnnotation/981755665</td>\n",
       "      <td>21083385</td>\n",
       "      <td>Genotypes AA + AG are associated with response to ivacaftor in people with Cystic Fibrosis.</td>\n",
       "      <td>0.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Clinical Annotation ID Evidence ID            Evidence Type  \\\n",
       "2              981755803   981755665  Variant Drug Annotation   \n",
       "\n",
       "                                           Evidence URL      PMID  \\\n",
       "2  https://www.pharmgkb.org/variantAnnotation/981755665  21083385   \n",
       "\n",
       "                                                                                       Summary  \\\n",
       "2  Genotypes AA + AG are associated with response to ivacaftor in people with Cystic Fibrosis.   \n",
       "\n",
       "  Score  \n",
       "2  0.25  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_ann_evidence[clinical_ann_evidence['Evidence ID'] == '981755665']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ed809c56-05af-4626-8efe-71662e83a694",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Genotype/Allele</th>\n",
       "      <th>Annotation Text</th>\n",
       "      <th>Allele Function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>981755803</td>\n",
       "      <td>AA</td>\n",
       "      <td>Patients with the rs75527207 AA genotype (two copies of the CFTR G551D variant) and cystic fibrosis may respond to ivacaftor treatment. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>981755803</td>\n",
       "      <td>AG</td>\n",
       "      <td>Patients with the rs75527207 AG genotype (one copy of the CFTR G551D variant) and cystic fibrosis may respond to ivacaftor treatment. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>981755803</td>\n",
       "      <td>GG</td>\n",
       "      <td>Patients with the rs75527207 GG genotype (do not have a copy of the CFTR G551D variant) and cystic fibrosis have an unknown response to ivacaftor treatment, as response may depend on the presence of other CFTR variants. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Clinical Annotation ID Genotype/Allele  \\\n",
       "0              981755803              AA   \n",
       "1              981755803              AG   \n",
       "2              981755803              GG   \n",
       "\n",
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Annotation Text  \\\n",
       "0                                                                                      Patients with the rs75527207 AA genotype (two copies of the CFTR G551D variant) and cystic fibrosis may respond to ivacaftor treatment. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.   \n",
       "1                                                                                        Patients with the rs75527207 AG genotype (one copy of the CFTR G551D variant) and cystic fibrosis may respond to ivacaftor treatment. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.   \n",
       "2  Patients with the rs75527207 GG genotype (do not have a copy of the CFTR G551D variant) and cystic fibrosis have an unknown response to ivacaftor treatment, as response may depend on the presence of other CFTR variants. FDA-approved drug labeling information and CPIC guidelines indicate use of ivacaftor in cystic fibrosis patients with at least one copy of a list of 33 CFTR genetic variants, including G551D. Other genetic and clinical factors may also influence response to ivacaftor.   \n",
       "\n",
       "  Allele Function  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_ann_alleles[clinical_ann_alleles[ID_COL_NAME] == '981755803']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2079c7d3-47c0-4016-8285-50208ed29dc8",
   "metadata": {},
   "source": [
    "Variant annotation sentence:\n",
    "```\n",
    "Genotypes AA + AG are associated with response to ivacaftor in people with Cystic Fibrosis.\n",
    "```\n",
    "Lines up exactly with clinical annotations which are on genotypes as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "3ffa4270-48da-4f97-a9ca-9084f9468d94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Evidence ID</th>\n",
       "      <th>Evidence Type</th>\n",
       "      <th>Evidence URL</th>\n",
       "      <th>PMID</th>\n",
       "      <th>Summary</th>\n",
       "      <th>Score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>981755803</td>\n",
       "      <td>982009991</td>\n",
       "      <td>Variant Drug Annotation</td>\n",
       "      <td>https://www.pharmgkb.org/variantAnnotation/982009991</td>\n",
       "      <td>23590265</td>\n",
       "      <td>Allele A is associated with response to ivacaftor in children with Cystic Fibrosis.</td>\n",
       "      <td>2.25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Clinical Annotation ID Evidence ID            Evidence Type  \\\n",
       "4              981755803   982009991  Variant Drug Annotation   \n",
       "\n",
       "                                           Evidence URL      PMID  \\\n",
       "4  https://www.pharmgkb.org/variantAnnotation/982009991  23590265   \n",
       "\n",
       "                                                                               Summary  \\\n",
       "4  Allele A is associated with response to ivacaftor in children with Cystic Fibrosis.   \n",
       "\n",
       "  Score  \n",
       "4  2.25  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_ann_evidence[clinical_ann_evidence['Evidence ID'] == '982009991']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5078781-6044-4fc4-9d8e-408c7c2b8573",
   "metadata": {},
   "source": [
    "On the other hand, this evidence is for the same clinical annotation but is for an allele."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9466024-7bb6-4fff-ae51-4ed0306b9c63",
   "metadata": {},
   "source": [
    "Allele types observed in variant annotation tables:\n",
    "* SNP genotype `C/T` or `CC` (annoying)\n",
    "* SNP allele `C`\n",
    "* insertion/deletion genotype `GGGGAGCTTTCCCAGAGACCC/del` (not sure about indel allele)\n",
    "* named allele `*17` or `HTTLPR short form (S allele)`\n",
    "* named genotype `*2/*4`\n",
    "* combinations of the above delineated with `+`\n",
    "* tricky ones that we'll ignore\n",
    "    * poor/intermediate/normal/ultrarapid metabolyzer, or combinations of these\n",
    "    * low/high activity\n",
    "    * slow/intermediate acetylator\n",
    "    * deficiency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f538be47-6014-4c27-b66e-1b9f089b2d6c",
   "metadata": {},
   "source": [
    "## Algorithm\n",
    "\n",
    "[Top of page](#Table-of-contents)\n",
    "\n",
    "#### Main\n",
    "1. First associate variant annotations with clinical annotations by ID\n",
    "2. For variant annotation, take the `Alleles` column\n",
    "3. Parse by splitting on `+` (and)\n",
    "4. For each piece, try to match exactly against the `Genotype/Allele` column in clinical annotation\n",
    "    * If it matches, go to \"Resolve combo\"\n",
    "5. Parse each piece by splitting on `/` to get alleles\n",
    "    * If no `/`, assume this to be an allele already\n",
    "6. Try to match against alleles parsed from clinical annotation (done as part of ID generation)\n",
    "    * If it matches, go to \"Resolve genotype\" and then \"Resolve combo\"\n",
    "7. If no match, log and exit\n",
    "\n",
    "#### Resolve combo (`+`)\n",
    "Variant annotation should be associated with each operand of `+`.\n",
    "\n",
    "#### Resolve genotype vs. allele\n",
    "We reach this state if we've matched something but the level isn't right - either variant annotation is per allele and clinical annotation is per genotype, or vice versa.\n",
    "\n",
    "**Case 1:** Variant annotation is per allele but clinical annotation is per genotype.\n",
    "* Variant annotation should be associated with each genotype containing the allele.\n",
    "\n",
    "**Case 2:** Variant annotation is per genotype but clinical annotation is per allele.\n",
    "* Variant annotation should be associated with each allele contained in its genotype."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f01a168d-283e-49e9-965f-9ba7eda37f50",
   "metadata": {},
   "source": [
    "Worked example 1: Variant annotation is for `*1/*18 + *2/*18 + *18/*18`, clinical annotations are on `*1`, `*2` and `*18`\n",
    "\n",
    "1. Break combo to `*1/*18` AND `*2/*18` AND `*18/*18`=> no match\n",
    "2. Split genotype to `*1` & `*18` AND `*2` & `*18` AND `*18` & `*18` => match\n",
    "3. Associate variant annotation with `*1`, `*2`, and `*18`\n",
    "\n",
    "Worked example 2: Variant annotation is on `AA + AG`, clinical annotations are on `AA`, `AG`, and `GG`.\n",
    "\n",
    "1. Break combo to `AA` and `AG` => match\n",
    "2. Associate variant annotation with `AA` and `AG`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a06a46f2-c575-4e5c-a544-edc3b9d39147",
   "metadata": {},
   "source": [
    "### Prototyping\n",
    "\n",
    "[Top of page](#Table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "id": "54b730ab-5449-4f4c-a81a-f3f8c13068bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from opentargets_pharmgkb.pandas_utils import split_and_explode_column\n",
    "import numpy as np\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 368,
   "id": "5d0a26be-3491-4ff3-8e38-f5d1f46bfea3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Examples from spreadsheet\n",
    "example_ca_ids = ['981755803', '1139506787', '1183888969', '1184514050', '981419266']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "id": "c1f9cbbf-1986-41e8-8c0c-cbb178ca09af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_evidence(caids=None):\n",
    "    if not caids:\n",
    "        caids = clinical_ann_evidence[ID_COL_NAME]\n",
    "    # Variant annotation ids\n",
    "    caid_to_vaid = {\n",
    "        caid: clinical_ann_evidence[clinical_ann_evidence[ID_COL_NAME] == caid]['Evidence ID'].to_list()\n",
    "        for caid in caids\n",
    "    }\n",
    "    return caid_to_vaid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "id": "ee4c339d-0c97-4534-88ed-b1bd4dd344cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extended_parse_genotype(genotype_string):\n",
    "    \"\"\"\n",
    "    Parse PGKB string representations of genotypes into alleles. Extended to include star alleles.\n",
    "    \"\"\"\n",
    "    alleles = [genotype_string]\n",
    "    \n",
    "    # SNPs\n",
    "    if len(genotype_string) == 2 and '*' not in genotype_string:\n",
    "        alleles = [genotype_string[0], genotype_string[1]]\n",
    "\n",
    "    # others\n",
    "    m = re.match('([^/]+)/([^/]+)', genotype_string, re.IGNORECASE)\n",
    "    if m:\n",
    "        alleles = [m.group(1), m.group(2)]\n",
    "\n",
    "    return alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "id": "8efba8da-418b-41de-91e4-32d65a534ce9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_associations_old(annotation_df, clinical_alleles_df):\n",
    "    \"\"\"\n",
    "    Older version for testing.\n",
    "    This was faster (no for-loop) and useful to see what variant annotation pieces don't get matched,\n",
    "    but some small bugs and not able to see which clinical annotations don't get matched.\n",
    "    \"\"\"\n",
    "    # Split on +\n",
    "    split_df = split_and_explode_column(annotation_df, 'Alleles', 'split_alleles_1', sep='\\+')\n",
    "    # Split on /\n",
    "    split_df = split_and_explode_column(split_df, 'split_alleles_1', 'split_alleles_2', sep='/')\n",
    "    # Get alleles from clinical annotations - same logic as for getting ids\n",
    "    split_clin_df = clinical_alleles_df.assign(parsed_genotype=clinical_alleles_df['Genotype/Allele'].apply(extended_parse_genotype))\n",
    "    split_clin_df = split_clin_df.explode('parsed_genotype').reset_index(drop=True)\n",
    "\n",
    "    # First match by the +-split alone\n",
    "    merged_df = pd.merge(split_clin_df, split_df, how='right', left_on='Genotype/Allele', right_on='split_alleles_1')\n",
    "\n",
    "    # Then try to fill in non-matches from /-split\n",
    "    merged_df_2 = pd.merge(split_clin_df, split_df, how='right', left_on='parsed_genotype', right_on='split_alleles_2')\n",
    "    merged_df.fillna(value={\n",
    "        'Clinical Annotation ID': merged_df_2['Clinical Annotation ID'],\n",
    "        'Genotype/Allele': merged_df_2['Genotype/Allele'],\n",
    "        'Annotation Text': merged_df_2['Annotation Text'],\n",
    "        'parsed_genotype': merged_df_2['parsed_genotype']\n",
    "    }, inplace=True)\n",
    "\n",
    "    # Select some columns to return & drop duplicates\n",
    "    return merged_df[\n",
    "    [ID_COL_NAME, 'Genotype/Allele', 'parsed_genotype', 'Annotation Text', 'Variant Annotation ID', 'Alleles', 'split_alleles_1', 'split_alleles_2', 'Sentence']\n",
    "    ].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "id": "48098ebc-6099-45db-89bf-325a1cc76bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_associations(annotation_df, clinical_alleles_df):\n",
    "    # Split on +\n",
    "    split_ann_df = split_and_explode_column(annotation_df, 'Alleles', 'split_alleles_1', sep='\\+')\n",
    "    # Split on /\n",
    "    split_ann_df = split_and_explode_column(split_ann_df, 'split_alleles_1', 'split_alleles_2', sep='/')\n",
    "    # Get alleles from clinical annotations - same logic as for getting ids\n",
    "    split_clin_df = clinical_alleles_df.assign(parsed_genotype=clinical_alleles_df['Genotype/Allele'].apply(extended_parse_genotype))\n",
    "    split_clin_df = split_clin_df.explode('parsed_genotype').reset_index(drop=True)\n",
    "\n",
    "    # Trim down columns \n",
    "    split_ann_df = split_ann_df[['Variant Annotation ID', 'Alleles', 'split_alleles_1', 'split_alleles_2', 'Sentence']]\n",
    "    split_clin_df = split_clin_df[[ID_COL_NAME, 'Genotype/Allele', 'parsed_genotype', 'Annotation Text']]\n",
    "\n",
    "    # Match by +-split and /-split\n",
    "    merged_df = pd.merge(split_clin_df, split_ann_df, how='outer', left_on='Genotype/Allele', right_on='split_alleles_1')\n",
    "    merged_df_2 = pd.merge(split_clin_df, split_ann_df, how='outer', left_on='parsed_genotype', right_on='split_alleles_2')\n",
    "\n",
    "    # If a genotype in a clinical annotation doesn't have evidence, want this listed with nan's\n",
    "    all_results = []\n",
    "    for _, genotype, parsed_genotype, _ in split_clin_df.itertuples(index=False):\n",
    "        # Rows that matched on genotype\n",
    "        rows_first_match = merged_df[(merged_df['Genotype/Allele'] == genotype) & (merged_df['parsed_genotype'] == parsed_genotype) & (~merged_df['Variant Annotation ID'].isna())]\n",
    "        # Rows that matched on parsed genotype\n",
    "        rows_second_match = merged_df_2[(merged_df_2['Genotype/Allele'] == genotype) & (merged_df_2['parsed_genotype'] == parsed_genotype) & (~merged_df_2['Variant Annotation ID'].isna())]\n",
    "        # If neither matches, add with nan's\n",
    "        if rows_first_match.empty and rows_second_match.empty:\n",
    "            all_results.append(merged_df[(merged_df['Genotype/Allele'] == genotype) & (merged_df['parsed_genotype'] == parsed_genotype)])\n",
    "        else:\n",
    "            all_results.extend([rows_first_match, rows_second_match])\n",
    "\n",
    "    final_result = pd.concat(all_results).drop_duplicates()\n",
    "\n",
    "    # If _no_ part of a variant annotation is associated with any clinical annotation, want this listed with nan's\n",
    "    for vaid, alleles, split_1, split_2, _ in split_ann_df.itertuples(index=False):\n",
    "        results_with_vaid = final_result[final_result['Variant Annotation ID'] == vaid]\n",
    "        if results_with_vaid.empty:\n",
    "            final_result = pd.concat((final_result,\n",
    "                                      merged_df[(merged_df['Variant Annotation ID'] == vaid) & (merged_df['Alleles'] == alleles) &\n",
    "                                                (merged_df['split_alleles_1'] == split_1) & (merged_df['split_alleles_2'] == split_2)]\n",
    "                                     ))\n",
    "    return final_result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 457,
   "id": "83c5ad18-d152-4fbd-ae8a-9f2509053371",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_results(caid_to_vaid):\n",
    "    results = {}\n",
    "    for caid, vaids in caid_to_vaid.items():\n",
    "        clinical_alleles_df = clinical_ann_alleles[clinical_ann_alleles[ID_COL_NAME] == caid]\n",
    "        # Get common columns among all types of variant annotations so we can concat\n",
    "        drug_df = var_drug_ann[var_drug_ann['Variant Annotation ID'].isin(vaids)][['Variant Annotation ID', 'Alleles', 'Sentence']]\n",
    "        phenotype_df = var_pheno_ann[var_pheno_ann['Variant Annotation ID'].isin(vaids)][['Variant Annotation ID', 'Alleles', 'Sentence']]\n",
    "        assay_df = var_fa_ann[var_fa_ann['Variant Annotation ID'].isin(vaids)][['Variant Annotation ID', 'Alleles', 'Sentence']]\n",
    "\n",
    "        all_variant_annotations = pd.concat((drug_df, phenotype_df, assay_df))\n",
    "        results[caid] = get_associations(all_variant_annotations, clinical_alleles_df)\n",
    "\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 458,
   "id": "c77d2b4b-a433-4dc8-910a-03bd962d9e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_results = get_results(get_evidence(example_ca_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 459,
   "id": "64025d53-d281-493d-a91f-e6ac07ae1c2c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "981755803: 62/62 (100.00%)\n",
      "1139506787: 168/171 (98.25%)\n",
      "1183888969: 12/12 (100.00%)\n",
      "1184514050: 39/39 (100.00%)\n",
      "981419266: 20/20 (100.00%)\n"
     ]
    }
   ],
   "source": [
    "# Approximate percentage of variant annotations we were able to associate to clinical annotation\n",
    "for caid in example_ca_ids:\n",
    "    num_rows = len(example_results[caid])\n",
    "    num_success = example_results[caid][ID_COL_NAME].count()\n",
    "    print(f'{caid}: {num_success}/{num_rows} ({num_success/num_rows*100:0.2f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "ecb9708d-9aff-419c-8e61-267813f93ff5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "981755803: 61/62 (98.39%)\n",
      "1139506787: 171/171 (100.00%)\n",
      "1183888969: 10/12 (83.33%)\n",
      "1184514050: 39/39 (100.00%)\n",
      "981419266: 20/20 (100.00%)\n"
     ]
    }
   ],
   "source": [
    "# Approximate percentage of clinical annotation genotype/alleles we were able to associate\n",
    "for caid in example_ca_ids:\n",
    "    num_rows = len(example_results[caid])\n",
    "    num_success = example_results[caid]['Variant Annotation ID'].count()\n",
    "    print(f'{caid}: {num_success}/{num_rows} ({num_success/num_rows*100:0.2f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "id": "a833c23e-e2f5-4787-a55c-9a632ab44521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Genotype/Allele</th>\n",
       "      <th>parsed_genotype</th>\n",
       "      <th>Annotation Text</th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>split_alleles_1</th>\n",
       "      <th>split_alleles_2</th>\n",
       "      <th>Sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1451663080</td>\n",
       "      <td>CT + TT</td>\n",
       "      <td>CT</td>\n",
       "      <td>CT</td>\n",
       "      <td>Genotypes CT + TT are associated with decreased metabolism of nicotine as compared to genotype CC.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1451663080</td>\n",
       "      <td>CT + TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>TT</td>\n",
       "      <td>Genotypes CT + TT are associated with decreased metabolism of nicotine as compared to genotype CC.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1451667320</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Allele T is associated with decreased metabolism of nicotine as compared to allele C.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1451672860</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>Allele T is associated with decreased metabolism of nicotine as compared to allele C.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1183703333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYP2A6 poor metabolizer is associated with decreased metabolism of nicotine in nonsmokers as compared to CYP2A6 normal metabolizer.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1183689160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYP2A6 poor metabolizer is associated with increased ratio of cotinine formation to removal when exposed to nicotine in nonsmokers as compared to CYP2A6 normal metabolizer.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1183689165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CYP2A6 poor metabolizer is associated with decreased ratio of plasma cotinine to urinary TNE when exposed to nicotine in smokers as compared to CYP2A6 normal metabolizer.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Clinical Annotation ID Genotype/Allele parsed_genotype Annotation Text  \\\n",
       "8                      NaN             NaN             NaN             NaN   \n",
       "9                      NaN             NaN             NaN             NaN   \n",
       "84                     NaN             NaN             NaN             NaN   \n",
       "85                     NaN             NaN             NaN             NaN   \n",
       "173                    NaN             NaN             NaN             NaN   \n",
       "10                     NaN             NaN             NaN             NaN   \n",
       "11                     NaN             NaN             NaN             NaN   \n",
       "\n",
       "    Variant Annotation ID  Alleles split_alleles_1 split_alleles_2  \\\n",
       "8              1451663080  CT + TT              CT              CT   \n",
       "9              1451663080  CT + TT              TT              TT   \n",
       "84             1451667320        T               T               T   \n",
       "85             1451672860        T               T               T   \n",
       "173            1183703333      NaN             NaN             NaN   \n",
       "10             1183689160      NaN             NaN             NaN   \n",
       "11             1183689165      NaN             NaN             NaN   \n",
       "\n",
       "                                                                                                                                                                         Sentence  \n",
       "8                                                                              Genotypes CT + TT are associated with decreased metabolism of nicotine as compared to genotype CC.  \n",
       "9                                                                              Genotypes CT + TT are associated with decreased metabolism of nicotine as compared to genotype CC.  \n",
       "84                                                                                          Allele T is associated with decreased metabolism of nicotine as compared to allele C.  \n",
       "85                                                                                          Allele T is associated with decreased metabolism of nicotine as compared to allele C.  \n",
       "173                                           CYP2A6 poor metabolizer is associated with decreased metabolism of nicotine in nonsmokers as compared to CYP2A6 normal metabolizer.  \n",
       "10   CYP2A6 poor metabolizer is associated with increased ratio of cotinine formation to removal when exposed to nicotine in nonsmokers as compared to CYP2A6 normal metabolizer.  \n",
       "11     CYP2A6 poor metabolizer is associated with decreased ratio of plasma cotinine to urinary TNE when exposed to nicotine in smokers as compared to CYP2A6 normal metabolizer.  "
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_results['1139506787'][example_results['1139506787'][ID_COL_NAME].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "id": "72917430-b993-424e-9f50-76785b823e7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Clinical Annotation ID</th>\n",
       "      <th>Genotype/Allele</th>\n",
       "      <th>parsed_genotype</th>\n",
       "      <th>Annotation Text</th>\n",
       "      <th>Variant Annotation ID</th>\n",
       "      <th>Alleles</th>\n",
       "      <th>split_alleles_1</th>\n",
       "      <th>split_alleles_2</th>\n",
       "      <th>Sentence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>981859108</td>\n",
       "      <td>*15:02 + *46:01</td>\n",
       "      <td>*46:01</td>\n",
       "      <td>*46:01</td>\n",
       "      <td>HLA-B *15:02 + *46:01 are associated with increased risk of Stevens-Johnson Syndrome when treated with phenytoin in Epilepsy.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Clinical Annotation ID Genotype/Allele parsed_genotype Annotation Text  \\\n",
       "18                    NaN             NaN             NaN             NaN   \n",
       "\n",
       "   Variant Annotation ID          Alleles split_alleles_1 split_alleles_2  \\\n",
       "18             981859108  *15:02 + *46:01          *46:01          *46:01   \n",
       "\n",
       "                                                                                                                         Sentence  \n",
       "18  HLA-B *15:02 + *46:01 are associated with increased risk of Stevens-Johnson Syndrome when treated with phenytoin in Epilepsy.  "
      ]
     },
     "execution_count": 372,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_results['981419266'][example_results['981419266'][ID_COL_NAME].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ed7e89e-dc47-4566-bad3-84af2cb7a502",
   "metadata": {},
   "source": [
    "7 missing from `1139506787`:\n",
    "* 4 are annotations on an RS that is part of the named haplotype\n",
    "* 3 are on \"poor metabolizers\"\n",
    "\n",
    "1 missing from `981419266`:\n",
    "* allele that is indeed not part of the clinical annotation for some reason\n",
    "\n",
    "Note these counts are after all the exploding, counting by variant annotation it should be something like 6 missing + 0 missing.\n",
    "\n",
    "Also this is counting coverage of variant annotations, not coverage of clinical annotations (e.g. not all genotypes in clincial annotations are present in these tables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 461,
   "id": "d6d0e173-581d-4e46-9e97-1f2805e4200b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for caid, table in example_results.items():\n",
    "    table.to_csv(f'{data_dir}/{caid}_association_test.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f38461a-1ab1-4c6a-af01-b712cd8161fb",
   "metadata": {},
   "source": [
    "### Full dataset run\n",
    "\n",
    "[Top of page](#Table-of-contents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "id": "27641722-c4c1-4a76-896e-79b14757869b",
   "metadata": {},
   "outputs": [],
   "source": [
    "full_results = get_results(get_evidence())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "91e1f67e-4e54-4b3b-9219-3eba55931efc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_variant_annotation_coverage(results):\n",
    "    \"\"\"Compute fraction of variant annotations that are covered by at least one clinical annotation genotype.\"\"\"\n",
    "    total_var_anns = 0\n",
    "    total_failure = 0\n",
    "    for caid in results:\n",
    "        var_anns = results[caid]['Variant Annotation ID'].dropna().unique()\n",
    "        total_var_anns += len(var_anns)\n",
    "        match = set(results[caid][~results[caid][ID_COL_NAME].isna()]['Variant Annotation ID'].dropna().unique())\n",
    "        non_match = set(results[caid][results[caid][ID_COL_NAME].isna()]['Variant Annotation ID'].dropna().unique())\n",
    "        total_failure += len(non_match - match)\n",
    "\n",
    "    total_success = total_var_anns - total_failure\n",
    "    print(f'Coverage of variant annotations: {total_success}/{total_var_anns} ({total_success/total_var_anns*100:0.2f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 499,
   "id": "b67cb58e-86fb-41cf-8b44-4b479cbc587c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_clinical_annotation_coverage(results):\n",
    "    \"\"\"Compute fraction of clinical annotation genotype/alleles that are covered by at least one variant annotation\"\"\"\n",
    "    total_genotypes = 0\n",
    "    total_failure = 0\n",
    "    \n",
    "    for caid in results:\n",
    "        genotypes = results[caid]['Genotype/Allele'].dropna().unique()\n",
    "        total_genotypes += len(genotypes)\n",
    "        match = set(results[caid][~results[caid]['Variant Annotation ID'].isna()]['Genotype/Allele'].dropna().unique())\n",
    "        non_match = set(results[caid][results[caid]['Variant Annotation ID'].isna()]['Genotype/Allele'].dropna().unique())\n",
    "        total_failure += len(non_match - match)\n",
    "        \n",
    "    total_success = total_genotypes - total_failure\n",
    "    print(f'Coverage of clinical annotation genotype/alleles: {total_success}/{total_genotypes} ({total_success/total_genotypes*100:0.2f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 505,
   "id": "5209136f-4ac1-473b-947e-df9e875c5c7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coverage of variant annotations: 13380/14280 (93.70%)\n"
     ]
    }
   ],
   "source": [
    "compute_variant_annotation_coverage(full_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 500,
   "id": "29b6fb60-16cb-42b5-9077-2127eb48f0ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Coverage of clinical annotation genotype/alleles: 10258/15755 (65.11%)\n"
     ]
    }
   ],
   "source": [
    "compute_clinical_annotation_coverage(full_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f423bcf-ec2c-4735-9f66-1618fccbaaf6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}